File: python/longest_substring_without_repeat_character.py
#!/usr/bin/env python

def find_longest_substring(raw):
    if not raw:
        return 0,None
    if len(raw) == 1:
        return 1,raw
    # i = left = cur_len = max_len = 0
    i, left, cur_len, max_len = 0, 0, 0, 0
    sub = ''
    collected_sub = {}
    max_sub = []
    char_dict = {}
    while i<len(raw):
        check_index = char_dict.get(raw[i])
        if check_index:
            if check_index < left:
                # sub += raw[i]
                sub = raw[left:i]
                cur_len += 1
            else:
                left = check_index + 1
                sub = raw[left:i]
                cur_len = i-left+1
        else:
            # sub += raw[i]
            sub = raw[left:i]
            cur_len += 1

        if cur_len >= max_len:
            max_len = cur_len
            collected_sub[sub] = max_len
        char_dict[raw[i]] = i
        i += 1

    for k,v in collected_sub.items():
        if v == max_len:
            max_sub.append(k)
    return max_len, max_sub

def main():
    raw_string = "abcdcafbkgzjaf"
    max_len,sub = find_longest_substring(raw_string)
    print "maximum length of longest substring is: %d, it's %s." % (max_len, sub)

if __name__ == "__main__":
    main()

File: golang/longest_substring_without_repeat_character.go
package main

import "fmt"

func find_longest_substring(raw string) int {

    left, max_len, cur_len := 0,0,0
    char_dict := map[byte]int{}
    for i:=0;i<len(raw);i++ {
        if char_dict[raw[i]] > 0 {
            if char_dict[raw[i]] < left {
                cur_len++
            } else {
                left = char_dict[raw[i]] + 1
                cur_len = i - char_dict[raw[i]] + 1
            }
        } else {
            cur_len++
        }

        if cur_len >= max_len {
            max_len = cur_len
        }
        char_dict[raw[i]] = i
    }
    return max_len
}


func main() {
    raw_string := "abcdcafbkgzjaf"
    max_len := find_longest_substring(raw_string)
    fmt.Printf("cur_len of longest substring is: %d\n" , max_len)
}
